docs/proposals/maintenance.md

```diff
@@ -0,0 +1,179 @@
+
+```
+type DrainSpec struct {
+  LameNodes    PodList
+  RemoveNodes  PodList
+  AddNodes     PodList
+  GracePeriod  int
+}
+```
+
+A drain object would work much like a deployment.  A drain controller would
+watch for drain objects and, within constraints, work toward the desired end
+state.  Nodes in `LameNodes` would merely be lamed, nodes is `RemoveNodes` would
+be drained according to the given parameters, and nodes in `AddNodes` would be
+undrained.  The object would be rejected if the sets overlap at all.  The action
+of the drain controller can be paused or stopped by deleting the drain object,
+and restarted by re-posting it.
+
+Undo would be trickier, but a clever client could be written such that it
+records the starting state of all the nodes and is able to send a reversing
+drain request.
+
+A central drain controller would have several advantages: safe restarts and the
+ability to react to other events in order to preserve the desired SLO.  For
+example, if a machine not in the list of RemoveNodes crashes and takes out other
+pods in replication sets also impacted by the drain, it can hold off on
+evictions impacting that replication set until MinSecondsBetweenEvictions and
+MinShardStrength are OK again.
+
+## Implementation of safety logic
+
+The implementation is the same whether there is one drain in flight or several.
+The drain agent (either kubectl or the drain controller) must watch for
+replication controllers and pods, and it will watch all pods to notice any time
+there is a disruption to a ReplicaSet.  Because there is no reliable mechanism
+to replay past events, when the controller restarts, it must wait until it
+builds up enough history to establish that it can enforce the safety
+constraints.  The standard loop on draining machines will be:
+
+- Is it safe to perform any evictions now?  If so, do them.
+- Are any machines empty?  If so mark them done.
+- Are any drains done?  If so mark them done.
+- Calculate the soonest we can do another eviction.
+- Sleep until then, or until we receive a watch update.
+
+## Future
+
+### Varying SLOs
+
+We could implement multiple SLOs, either by allowing the cell admin to specify
+choices or by allowing this to be specified in the ReplicaSet or referring to
+the similar parameters already present in Deployment.  I believe we should offer
+this safety to pods regardless of whether they're managed by Deployment or not,
+so I don't think Deployment is the right place to pull this from.  Rather, when
+a sketch for multiple or parameterized SLOs as a feature is present, we should
+discuss representation that would allow a set of parameters to apply to any
+ReplicaSet, or perhaps any arbitrary collection of Pods matched by a
+selector.
+
+### More Sophisticated SLO Expression
+
+Rather than only waiting `X` seconds between evictions, an admin might want to
+limit the total in, say, the trailing 6 or 24 hour window.  This requires more
+reliable event logging, described below.
```

Comment:
@erictune I don't understand. Pods may have been deleted so I think there's not enough information.